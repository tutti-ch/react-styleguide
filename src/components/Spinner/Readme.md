Dark spinner:

```jsx
<div style={{ textAlign: 'center', padding: '1.5%' }}>
    <Spinner size={Spinner.SIZE_SMALL} color={Spinner.COLOR_DARK} />
    <Spinner size={Spinner.SIZE_MEDIUM} color={Spinner.COLOR_DARK} />
    <Spinner size={Spinner.SIZE_LARGE} color={Spinner.COLOR_DARK} />
    <Spinner size={11} color={Spinner.COLOR_DARK} />
</div>
```

Light spinner:

```jsx
<div style={{ backgroundColor: '#dbdbdb', padding: '1.5%', textAlign: 'center' }}>
  <Spinner size={Spinner.SIZE_SMALL} color={Spinner.COLOR_LIGHT} />
  <Spinner size={Spinner.SIZE_MEDIUM} color={Spinner.COLOR_LIGHT} />
  <Spinner size={Spinner.SIZE_LARGE} color={Spinner.COLOR_LIGHT} />
  <Spinner size={11} color={Spinner.COLOR_LIGHT} />
</div>
```
