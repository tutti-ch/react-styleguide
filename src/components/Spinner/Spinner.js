import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classes from './Spinner.scss'
import classNames from 'classnames'

/**
 * The spinner component creates a spinning div to indicate
 * the user that the content is loading.
 */
export default class Spinner extends Component {
  static COLOR_LIGHT = 'light'
  static COLOR_DARK = 'dark'
  static SIZE_SMALL = 2
  static SIZE_MEDIUM = 5
  static SIZE_LARGE = 8

  static propTypes = {
    /**
     * Define the color of the spinner.
     */
    color: PropTypes.oneOf([Spinner.COLOR_DARK, Spinner.COLOR_LIGHT]),

    /**
     * The size of the spinner. This will determine the thickness of the spinner.
     */
    size: PropTypes.number,

    /**
     * The additional class name that will be included to handle specific cases.
     */
    className: PropTypes.string
  }

  static defaultProps = {
    size: Spinner.SIZE_MEDIUM,
    color: Spinner.COLOR_LIGHT
  }

  constructor(props) {
    super(props)

    this.calculateStyles = this.calculateStyles.bind(this)
  }

  /**
   * Calculate the styles that will be injected to the spinner.
   */
  calculateStyles() {
    const { color, size } = this.props

    let rgba, intensity

    if (color === Spinner.COLOR_DARK) {
      rgba = 0
      intensity = 0.1
    } else {
      rgba = 255
      intensity = 0.6
    }

    const thickness = size / 10
    const border = `${thickness}rem solid rgba(${rgba}, ${rgba}, ${rgba}, ${intensity})`
    const sizeInRem = `${size}rem`

    return {
      width: sizeInRem,
      height: sizeInRem,
      borderTop: border,
      borderRight: border,
      borderBottom: border,
      borderLeft: `${thickness}rem solid rgba(${rgba}, ${rgba}, ${rgba}, ${Math.round((intensity + 0.2) * 100) / 100})`
    }
  }

  /**
   * Render the spinner.
   */
  render() {
    const { className } = this.props
    const styles = this.calculateStyles()

    return (
      <div className={classNames(classes.wrapper, className)}>
        <div className={classes.spinner} style={styles}/>
      </div>
    )
  }
}
