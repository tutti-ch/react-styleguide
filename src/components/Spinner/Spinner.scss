.wrapper {
  display: inline-block;
}

.spinner {
  transform: translateZ(0);
  animation: spinMe 0.8s infinite linear;
  display: inline-block;
}

.spinner {
  border-radius: 50%;
}

.spinner:after {
  border-radius: 50%;
  height: 100%;
  width: 100%;
}

@keyframes spinMe {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
