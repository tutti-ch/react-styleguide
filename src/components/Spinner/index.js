export { default } from "./Spinner.js";
