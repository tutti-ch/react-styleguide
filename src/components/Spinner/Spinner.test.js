/* global describe, test */
import React from 'react'
import { shallow, mount } from 'enzyme'
import Spinner from './index'
import toJson from 'enzyme-to-json'

describe('(Component) Spinner', () => {
  test('should calculate the styles correctly', () => {
    const comp = shallow(<Spinner color={Spinner.COLOR_LIGHT} />)
    const inst = comp.instance()

    expect(inst.calculateStyles()).toEqual({
      borderBottom: '0.5rem solid rgba(255, 255, 255, 0.6)',
      borderLeft: '0.5rem solid rgba(255, 255, 255, 0.8)',
      borderRight: '0.5rem solid rgba(255, 255, 255, 0.6)',
      borderTop: '0.5rem solid rgba(255, 255, 255, 0.6)',
      height: '5rem',
      width: '5rem'
    })

    comp.setProps({ color: Spinner.COLOR_DARK, size: Spinner.SIZE_SMALL })

    expect(inst.calculateStyles()).toEqual({
      borderBottom: '0.2rem solid rgba(0, 0, 0, 0.1)',
      borderLeft: '0.2rem solid rgba(0, 0, 0, 0.3)',
      borderRight: '0.2rem solid rgba(0, 0, 0, 0.1)',
      borderTop: '0.2rem solid rgba(0, 0, 0, 0.1)',
      height: '2rem',
      width: '2rem'
    })
  })

  test('should match the snapshot', () => {
    const comp = mount(<Spinner size={Spinner.SIZE_LARGE} color={Spinner.COLOR_DARK} className='my-class'/>)
    expect(toJson(comp)).toMatchSnapshot()
  })
})