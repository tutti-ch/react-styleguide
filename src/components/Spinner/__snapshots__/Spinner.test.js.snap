// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Spinner should match the snapshot 1`] = `
<Spinner
  className="my-class"
  color="dark"
  size={8}
>
  <div
    className="my-class"
  >
    <div
      className={undefined}
      style={
        Object {
          "borderBottom": "0.8rem solid rgba(0, 0, 0, 0.1)",
          "borderLeft": "0.8rem solid rgba(0, 0, 0, 0.3)",
          "borderRight": "0.8rem solid rgba(0, 0, 0, 0.1)",
          "borderTop": "0.8rem solid rgba(0, 0, 0, 0.1)",
          "height": "8rem",
          "width": "8rem",
        }
      }
    />
  </div>
</Spinner>
`;
